pkg/controller/garbagecollector/garbagecollector.go

```diff
@@ -0,0 +1,218 @@
+/*
+Copyright 2016 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package gabagecollector
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/golang/glog"
+
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/errors"
+	"k8s.io/kubernetes/pkg/api/meta"
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	"k8s.io/kubernetes/pkg/api/v1"
+	"k8s.io/kubernetes/pkg/apimachinery/registered"
+	"k8s.io/kubernetes/pkg/client/cache"
+	"k8s.io/kubernetes/pkg/client/typed/dynamic"
+	"k8s.io/kubernetes/pkg/controller/framework"
+	"k8s.io/kubernetes/pkg/runtime"
+	"k8s.io/kubernetes/pkg/util/wait"
+	"k8s.io/kubernetes/pkg/util/workqueue"
+	"k8s.io/kubernetes/pkg/watch"
+)
+
+const ResourceResyncTime = 30 * time.Second
+
+type GarbageCollector struct {
+	restMapper         meta.RESTMapper
+	clientPool         dynamic.ClientPool
+	dirty_queue        *workqueue.Type
+	scannerStores      []cache.Store
+	scannerControllers []*framework.Controller
+}
+
+func NewGarbageCollector(clientPool dynamic.ClientPool, resources []unversioned.GroupVersionResource) *GarbageCollector {
+	gc := &GarbageCollector{
+		clientPool:  clientPool,
+		dirty_queue: workqueue.New(),
+		// TODO: should use a dynamic RESTMapper built from the discovery results.
+		restMapper: registered.RESTMapper(),
+	}
+	for _, resource := range resources {
```

Comment:
> So jobs under extensions/v1beta1 would indicate they were not managed by GC, even though you could set owner references and finalizers on them, and GC would manage those jobs when they showed up under batch/v1? That seems really weird.

Yeah, that's weird. How about adding a `APIResource.CanonicalGCResource`, which refers to the canonical version of the resource that is managed by GC? And if the field is empty, then GC won't manage the resource.

Or, we can add two fields, `APIResource.CanonicalResource` referring the canonical version of a resource, and `APIResource.ManagedByGC` indicating if GC should manage this resource.